### ISSUE TITLE
pwndbg does not work on qemu-system-aarch64 after mmu is enabled

### ISSUE THREAD

**Issue by kooWZ** [pwndbg/pwndbg#2472]
### Description
I am using pwndbg for a tiny OS running on qemu-system-aarch64.
I encountered two problems:
1. pwndbg shows encoding conversion problem and "ERROR: Could not find ELF base!"
```log
Reading symbols from /root/kernel8.elf...
warning: could not convert 'rb_root' from the host encoding (UTF-8) to UTF-32.
This normally should not happen, please file a bug report.
Remote debugging using localhost:1234
0x0000000040000000 in ?? ()
ERROR: Could not find ELF base!
```

2. Here is part of my `_start` function:
```asm
  /* Enable MMU. */
  mrs x9, sctlr_el1
  orr x9, x9, #SCTLR_MMU_ENABLED
  msr sctlr_el1, x9
```
pwndbg single step runs well before the last line `msr sctlr_el1, x9`. After that, pwndbg shows the following error and loses control:
```log
*PC   0x40000028
───────────────────────────────────────────────────────────────────────────────[ DISASM / aarch64 / set emulate on ]────────────────────────────────────────────────────────────────────────────────
Invalid address 0x40000028










─────────────────────────────────────────────────────────────────────────────────────────────[ STACK ]──────────────────────────────────────────────────────────────────────────────────────────────
<Could not read memory at 0x0>
───────────────────────────────────────────────────────────────────────────────────────────[ BACKTRACE ]────────────────────────────────────────────────────────────────────────────────────────────
 ► 0       0x40000028
   1              0x0
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```
However, vanilla GDB works well without errors.

### Steps to reproduce
1. Compile the OS into a elf file
2. Run QEMU with: 
```bash
qemu-system-aarch64 \
    -machine virt,gic-version=3 \
    -cpu cortex-a72 \
    -smp 1 \
    -m 4096 \
    -nographic \
    -monitor none \
    -serial "mon:stdio" \
    -global virtio-mmio.force-legacy=false \
    -kernel "${kernel_elf}" \
    -gdb tcp::1234 -S
```
3. Run pwndbg with:
```bash
pwndbg \
  -ex "set architecture aarch64" \
  -ex "file ${kernel_elf}" \
  -ex "target remote localhost:1234" 
```

### My setup
pwndbg version:
```log
Gdb:      15.1
Python:   3.12.4 (main, Jun  6 2024, 18:26:44) [GCC 13.3.0]
Pwndbg:   2024.08.29
Capstone: 5.0.1280
Unicorn:  2.0.1
```
QEMU version:
```log
QEMU emulator version 8.2.2 (Debian 1:8.2.2+ds-0ubuntu1.2)
```
Output of `file /root/kernel8.elf`:
```bash
ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, with debug_info, not stripped
```

**Comment by disconnect3d** [comment]
> However, vanilla GDB works well without errors.

What does this mean? Can vanilla GDB display assembly when you do `p/10i $pc`? The context of Pwndbg literally reads the bytes from memory and disassembles them with capstone, so if it doesn't display, its likely that GDB just can't read that part of memory either.

So its most likely not that Pwndbg doesnt work, but your tiny OS kernel doesn't provide what it should for the MMU/QEMU GDBstub to work properly. Another option could be that there is some QEMU kernel bug/unhandled situation.

**Comment by kooWZ** [comment]
> > However, vanilla GDB works well without errors.
> 
> What does this mean? 

My OS starts at `_start`. After enabling MMU and doing other configurations, it calls `main`.
When using vanilla GDB:
- If single step at the beginning, it raises error "Cannot find bounds of current function"
- If `b main` then `c`, it stops at `main` successfully. 

When using pwndbg:
- Single step at the beginning is successful until MMU is enabled.
- If `b main` then `c`, it fails to stop at `main`.

**Comment by disconnect3d** [comment]
Hmm...

> After that, pwndbg shows the following error and loses control

Does this mean that further `si` do not advance/change the `PC` register?

I doubt it very very much, but can you try using a `memoize` command and single step at this point? This command disables caching and it could be some bug with us caching stuff too much which would make us display an old context.

**Comment by disconnect3d** [comment]
@kooWZ btw can u share ur tiny kernel binary so that we could reproduce this if we ever wanted to debug it ourselves? (I am not sure if I get to find time for that, but maybe someone else would tackle it upon).

(Ideally, provide a repo we could build? For those who wants to try it, be aware that its untrusted code :p)

**Comment by kooWZ** [comment]
@disconnect3d I think I found the reason, but still I have totally no idea why this would happen:
When attaching QEMU with vanilla GDB and pwndbg, a certain instruction in memory changes!

I use the same command to launch QEMU,  also I use the same arguments to launch gdb/pwndbg.
My assembly code:
```asm
  adrp x9, kernel_pt_level0
  msr ttbr0_el1, x9
  msr ttbr1_el1, x9
```
Vanilla GDB:
![image](https://github.com/user-attachments/assets/62f035f1-b161-40e6-92d0-31e205d3076b)
Pwndbg:
![image](https://github.com/user-attachments/assets/dc6bc036-841a-4eac-81ed-9d11a9ab7116)

The first line of code changed, making the configuation of MMU failed, so pwndbg cannot read from memory after MMU is enabled.

I think this is a really weird bug. Do you have any idea if this may be related to pwndbg?

> @kooWZ btw can u share ur tiny kernel binary so that we could reproduce this if we ever wanted to debug it ourselves? 

This OS is part of a university course assignment, so it may not be convenient to share publicly. If you think the code or elf file would be helpful for debugging, I might be able to send them privately.

**Comment by kooWZ** [comment]
@disconnect3d I figured out this problem was caused by some kind of encoding problem in `poke` function in `memory.py`.
It tries to check whether `0x40000000` is writeable.
https://github.com/pwndbg/pwndbg/blob/f945b417da14c5bafdc41e6c0c38d83100501e0d/pwndbg/gdblib/memory.py#L114-L130
https://github.com/pwndbg/pwndbg/blob/f945b417da14c5bafdc41e6c0c38d83100501e0d/pwndbg/gdblib/memory.py#L149-L171
First it reads a byte from the addr using `peek`. In my case it got a byte `0xa9`, returned as a str `c`.
Then it calls `write` to write this byte back to the address. 
However in `write`, there is a encoding procedure
https://github.com/pwndbg/pwndbg/blob/f945b417da14c5bafdc41e6c0c38d83100501e0d/pwndbg/gdblib/memory.py#L107-L108
After that `c` is encoded as `\xc2\xa9`, and these two bytes were written to memory, changing the instruction.

**Comment by kooWZ** [comment]
I modified 
https://github.com/pwndbg/pwndbg/blob/f945b417da14c5bafdc41e6c0c38d83100501e0d/pwndbg/gdblib/memory.py#L107-L108
to:
```py
    if isinstance(data, str):
        try:
            data = bytes(data, "latin1")
        except:
            data = bytes(data, "utf-8")
```
and now pwndbg works fluently.

**Comment by disconnect3d** [comment]
Nice investigation! I am surprised we didn’t fall into this bug before, or,
we did but we did not notice.

Anyway, I think there is no reason for peek to return a str instead of the
bytes type, so we should probably change the code accordingly. This way we
would not have to do the additional encoding in the write (also, shouldn’t
we just use utf8 all the time instead of Latin? Huh).

**Comment by OBarronCS** [comment]
Wow! Great find - also very surprised this hasn't been noticed before.

Another fix for this would be to remove the `chr` iand `[0]` n this line in `peek` https://github.com/pwndbg/pwndbg/blob/f945b417da14c5bafdc41e6c0c38d83100501e0d/pwndbg/gdblib/memory.py#L127 and replace it with `bytes` so it's `bytes(read(address, 1))`. Would need to change some associated type hints and go through all the places that call peek to make sure bytes works (there's also a mirrored function in `agblib/memory.py` that would need the same change)

**Comment by disconnect3d** [comment]
@OBarronCS yeah this is what we have to do: change peek to return a bytearray. That's what my PR does but I implemented too much and uncovered some bugs we need to fix as well

**Comment by disconnect3d** [comment]
Eh, I am a bit out of time recently. The immediate fix for this is pretty trivial and if someone wants to send it in a PR before me, please do so! Part of it is implemented in #2477 but I went too far with changing the exception cases...

@kooWZ any chance you can send a PR fixing this immediate issue?

@OBarronCS do you maybe have time to go with a more broad fix for the exception case? (As we spoke part of the code that reaches the exception lies in disasm code which enhances display :x)

**Comment by disconnect3d** [comment]
The main issue here has been fixed in #2483. The other issue of us accepting bogus addresses for `memory.{peek,poke}` is not addressed yet.
